flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/ChangelogHandleReaderWithCache.java

```diff
@@ -0,0 +1,184 @@
+        this.cacheFileCreator = RefCountedTmpFileCreator.inDirectories(tempFiles);
+    }
+
+    @Override
+    public DataInputStream apply(StreamStateHandle handle, Long offset) throws IOException {
+        if (!(handle instanceof FileStateHandle)) {
+            return wrapAndSeek(handle.openInputStream(), offset);
+        }
+
+        FileStateHandle fileHandle = (FileStateHandle) handle;
+        DataInputStream input;
+
+        if (fileHandle.getFilePath().getFileSystem().isDistributedFS()) {
+
+            Path dfsPath = fileHandle.getFilePath();
+
+            final RefCountedBufferingFileStream refCountedFileStream =
+                    cache.computeIfAbsent(
+                            dfsPath,
+                            key -> {
+                                RefCountedBufferingFileStream fileStream = null;
+                                FSDataInputStream handleInputStream = null;
+
+                                try {
+                                    fileStream =
+                                            RefCountedBufferingFileStream.openNew(cacheFileCreator);
+                                    handleInputStream = handle.openInputStream();
+                                    IOUtils.copyBytes(wrap(handleInputStream), fileStream);
+                                    LOG.debug(
+                                            "download and decompress dstl file : {} to cached file : {}",
+                                            fileHandle.getFilePath(),
+                                            fileStream.getInputFile().getPath());
+                                } catch (IOException e) {
+                                    ExceptionUtils.rethrow(e);
+                                } finally {
+                                    if (handleInputStream != null) {
+                                        IOUtils.closeQuietly(handleInputStream);
+                                    }
+                                }
+
+                                return fileStream;
+                            });
+
+            FileInputStream fin = null;
+            synchronized (cache) {
+                if (refCountedFileStream.getReferenceCounter() >= 1) {
+                    fin = new FileInputStream(refCountedFileStream.getInputFile());
+                    if (offset != 0) {
+                        LOG.debug("seek to {}", offset);
+                        fin.getChannel().position(offset);
+                    }
+                    refCountedFileStream.retain();
+                } else {
+                    // cache file already be deleted, try again
+                    return apply(handle, offset);
+                }
+            }
+
+            input =
+                    new DataInputStream(new BufferedInputStream(fin)) {
+                        @Override
+                        public void close() throws IOException {
+                            super.close();
```

Comment:
Should it be wrapped into `try` to guarantee decrementing the RC?

Reply:
Yes, you are right.